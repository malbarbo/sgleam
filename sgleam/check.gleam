@external(javascript, "../sgleam/sgleam_ffi.mjs", "check_equal")
pub fn eq(a: x, b: x) -> Bool

@external(javascript, "../sgleam/sgleam_ffi.mjs", "check_approx")
pub fn approx(a: Float, b: Float, tolerance: Float) -> Bool

@external(javascript, "../sgleam/sgleam_ffi.mjs", "check_true")
pub fn true(val: Bool) -> Bool

@external(javascript, "../sgleam/sgleam_ffi.mjs", "check_false")
pub fn false(val: Bool) -> Bool
