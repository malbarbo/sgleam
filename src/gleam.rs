use camino::{Utf8Path, Utf8PathBuf};
use gleam_core::{
    build::{
        Mode, Module, NullTelemetry, PackageCompiler, StaleTracker, Target,
        TargetCodegenConfiguration, Telemetry,
    },
    config::PackageConfig,
    error::{FileIoAction, FileKind},
    io::{memory::InMemoryFileSystem, FileSystemWriter},
    javascript::is_bigint_enabled,
    parse::parse_module,
    type_::{printer::Printer, Type},
    uid::UniqueIdGenerator,
    warning::{VectorWarningEmitterIO, WarningEmitter, WarningEmitterIO},
    Error, Warning,
};
use std::{
    collections::{HashSet, VecDeque},
    io::{Read, Write},
    path::PathBuf,
    rc::Rc,
    sync::Arc,
    time::{Duration, Instant, SystemTime},
};
use tar::Archive;
use termcolor::{Color, ColorSpec, WriteColor};

use crate::{error::stderr_buffer_writer, GLEAM_STDLIB, GLEAM_STDLIB_BIGINT};

#[derive(Clone)]
pub struct Project {
    pub fs: InMemoryFileSystem,
}

fn stdlib() -> &'static [u8] {
    if is_bigint_enabled() {
        GLEAM_STDLIB_BIGINT
    } else {
        GLEAM_STDLIB
    }
}

impl Default for Project {
    fn default() -> Project {
        let mut project = Project {
            fs: InMemoryFileSystem::new(),
        };

        extract_tar(&mut project.fs, Archive::new(stdlib()), Project::source())
            .expect("Extract stdlib");
        project.write_source("sgleam/check.gleam", crate::SGLEAM_CHECK);
        project.write_source("sgleam_ffi.mjs", crate::SGLEAM_FFI_MJS);
        project.write_out("prelude.mjs", gleam_core::javascript::prelude());
        project
    }
}

impl Project {
    pub fn root() -> &'static Utf8Path {
        "/".into()
    }

    pub fn source() -> &'static Utf8Path {
        "/src".into()
    }

    pub fn out() -> &'static Utf8Path {
        "/build".into()
    }

    pub fn prelude() -> &'static Utf8Path {
        "/build/prelude.mjs".into()
    }

    pub fn write_source(&mut self, name: &str, content: &str) {
        let path = Project::source().join(name);
        self.fs
            .write(&path, content)
            .expect("Write a file in memory");
        self.fs
            .try_set_modification_time(&path, SystemTime::now())
            .expect("Set modification time of a file in memory")
    }

    pub fn copy_file_to_source(&mut self, input: &Utf8Path) -> Result<(), Error> {
        let content = std::fs::read_to_string(input).map_err(|err| Error::FileIo {
            kind: FileKind::File,
            action: FileIoAction::Read,
            path: input.into(),
            err: Some(err.to_string()),
        })?;
        self.write_source(input.as_str(), &content);
        Ok(())
    }

    pub fn write_out(&mut self, name: &str, content: &str) {
        let path = Project::out().join(name);
        self.fs
            .write(&path, content)
            .expect("Write a file in memory");
    }
}

pub fn get_module<'a>(modules: &'a [Module], name: &str) -> Option<&'a Module> {
    modules.iter().find(|m| m.name == name)
}

// FIXME: check in the lsp module how the action "Add type annotation" works
pub fn type_to_string(module: &Module, type_: &Type) -> String {
    Printer::new(&module.ast.names).print_type(type_).into()
}

pub fn fn_type_to_string(module: &Module, args: &[Arc<Type>], retrn: Arc<Type>) -> String {
    type_to_string(
        module,
        &Type::Fn {
            args: args.into(),
            retrn,
        },
    )
}

// TODO: move this function to Project
pub fn compile(project: &mut Project, repl: bool) -> Result<Vec<Module>, Error> {
    let config = PackageConfig {
        target: Target::JavaScript,
        ..Default::default()
    };

    let target = TargetCodegenConfiguration::JavaScript {
        emit_typescript_definitions: false,
        prelude_location: Project::prelude().into(),
    };

    let mut compiler = PackageCompiler::new(
        &config,
        Mode::Dev,
        Project::root(),
        Project::out(),
        Project::out(),
        &target,
        UniqueIdGenerator::new(),
        project.fs.clone(),
    );

    compiler.write_metadata = true;

    compiler
        .compile(
            &WarningEmitter::new(Rc::new(ConsoleWarningEmitter::with_repl(repl))),
            &mut im::HashMap::new(),
            &mut im::HashMap::new(),
            &mut StaleTracker::default(),
            &mut HashSet::new(),
            &NullTelemetry,
        )
        .map(|out| out.modules)
        .into_result()
}

pub fn find_imports(paths: Vec<Utf8PathBuf>) -> Result<Vec<Utf8PathBuf>, gleam_core::Error> {
    let warning_emitter = WarningEmitter::new(Rc::new(VectorWarningEmitterIO::new()));
    let mut files: Vec<Utf8PathBuf> = vec![];
    let mut pending = VecDeque::from(paths);
    while let Some(path) = pending.pop_front() {
        if files.contains(&path) {
            continue;
        }

        files.push(path.clone());

        let src = std::fs::read_to_string(&path).map_err(|err| gleam_core::Error::FileIo {
            kind: FileKind::File,
            action: FileIoAction::Read,
            path: path.clone(),
            err: Some(err.to_string()),
        })?;

        let parsed = parse_module(path.clone(), &src, &warning_emitter).map_err(|error| {
            gleam_core::Error::Parse {
                path,
                src: src.into(),
                error,
            }
        })?;

        for definition in &parsed.module.definitions {
            match &definition.definition {
                gleam_core::ast::Definition::Import(import)
                    if import.module != "sgleam"
                        && !import.module.starts_with("sgleam/")
                        && !import.module.starts_with("gleam/") =>
                {
                    let mut path = Utf8PathBuf::new();
                    for p in import.module.split("/") {
                        path.push(p);
                    }
                    path.set_extension("gleam");
                    pending.push_back(path);
                }
                _ => continue,
            }
        }
    }
    Ok(files)
}

fn extract_tar(
    fs: &mut InMemoryFileSystem,
    mut arch: Archive<&[u8]>,
    to: &Utf8Path,
) -> Result<(), Error> {
    let mut buf = vec![];
    for entry in arch.entries().map_err(to_error_stdio)? {
        let mut entry = entry.map_err(to_error_stdio)?;
        let is_dir = entry.header().entry_type().is_dir();
        let entry_path = entry.path().map_err(to_error_stdio)?.into_owned();
        let entry_path = Utf8PathBuf::from_path_buf(entry_path).map_err(to_error_nonutf8_path)?;
        let path = to.join(entry_path);
        if is_dir {
            fs.mkdir(&path)?;
        } else {
            buf.clear();
            entry.read_to_end(&mut buf).map_err(to_error_stdio)?;
            fs.write_bytes(&path, &buf)?;
        }
    }
    Ok(())
}

fn to_error_stdio(err: std::io::Error) -> Error {
    Error::StandardIo {
        action: gleam_core::error::StandardIoAction::Read,
        err: Some(err.kind()),
    }
}

pub fn to_error_nonutf8_path(path: PathBuf) -> Error {
    Error::NonUtf8Path { path }
}

// The remaining of this file is copied from gleam project

#[derive(Debug, Default, Clone)]
pub struct Reporter;

impl Reporter {
    pub fn new() -> Self {
        Self
    }
}

impl Telemetry for Reporter {
    fn compiled_package(&self, duration: Duration) {
        print_compiled(duration);
    }

    fn compiling_package(&self, name: &str) {
        print_compiling(name);
    }

    fn checked_package(&self, duration: Duration) {
        print_checked(duration);
    }

    fn checking_package(&self, name: &str) {
        print_checking(name);
    }

    fn downloading_package(&self, name: &str) {
        print_downloading(name)
    }

    fn packages_downloaded(&self, start: Instant, count: usize) {
        print_packages_downloaded(start, count)
    }

    fn resolving_package_versions(&self) {
        print_resolving_versions()
    }

    fn running(&self, name: &str) {
        print_running(name);
    }

    fn waiting_for_build_directory_lock(&self) {
        print_waiting_for_build_directory_lock()
    }
}

pub fn print_published(duration: Duration) {
    print_colourful_prefix("Published", &format!("in {}", seconds(duration)))
}

pub fn print_retired(package: &str, version: &str) {
    print_colourful_prefix("Retired", &format!("{package} {version}"))
}

pub fn print_unretired(package: &str, version: &str) {
    print_colourful_prefix("Unretired", &format!("{package} {version}"))
}

pub fn print_publishing_documentation() {
    print_colourful_prefix("Publishing", "documentation");
}

fn print_downloading(text: &str) {
    print_colourful_prefix("Downloading", text)
}

fn print_waiting_for_build_directory_lock() {
    print_colourful_prefix("Waiting", "for build directory lock")
}

fn print_resolving_versions() {
    print_colourful_prefix("Resolving", "versions")
}

fn print_compiling(text: &str) {
    print_colourful_prefix("Compiling", text)
}

pub(crate) fn print_checking(text: &str) {
    print_colourful_prefix("Checking", text)
}

pub(crate) fn print_compiled(duration: Duration) {
    print_colourful_prefix("Compiled", &format!("in {}", seconds(duration)))
}

pub(crate) fn print_checked(duration: Duration) {
    print_colourful_prefix("Checked", &format!("in {}", seconds(duration)))
}

pub(crate) fn print_running(text: &str) {
    print_colourful_prefix("Running", text)
}

fn print_packages_downloaded(start: Instant, count: usize) {
    let elapsed = seconds(start.elapsed());
    let msg = match count {
        1 => format!("1 package in {elapsed}"),
        _ => format!("{count} packages in {elapsed}"),
    };
    print_colourful_prefix("Downloaded", &msg)
}

pub fn seconds(duration: Duration) -> String {
    format!("{:.2}s", duration.as_millis() as f32 / 1000.)
}

pub fn print_colourful_prefix(prefix: &str, text: &str) {
    let buffer_writer = stderr_buffer_writer();
    let mut buffer = buffer_writer.buffer();
    buffer
        .set_color(
            ColorSpec::new()
                .set_intense(true)
                .set_fg(Some(Color::Magenta)),
        )
        .expect("print_colourful_prefix");
    write!(buffer, "{prefix: >11}").expect("print_colourful_prefix");
    buffer
        .set_color(&ColorSpec::new())
        .expect("print_colourful_prefix");
    writeln!(buffer, " {text}").expect("print_colourful_prefix");
    buffer_writer
        .print(&buffer)
        .expect("print_colourful_prefix");
}

#[derive(Debug, Clone, Copy)]
pub struct ConsoleWarningEmitter {
    repl: bool,
}

impl ConsoleWarningEmitter {
    pub fn with_repl(repl: bool) -> ConsoleWarningEmitter {
        ConsoleWarningEmitter { repl }
    }
}

impl WarningEmitterIO for ConsoleWarningEmitter {
    fn emit_warning(&self, warning: Warning) {
        if self.repl {
            if let Warning::Type {
                warning:
                    gleam_core::type_::Warning::Todo { .. }
                    | gleam_core::type_::Warning::UnreachableCodeAfterPanic { .. }
                    | gleam_core::type_::Warning::UnusedConstructor { .. }
                    | gleam_core::type_::Warning::UnusedImportedModule { .. }
                    | gleam_core::type_::Warning::UnusedImportedModuleAlias { .. }
                    | gleam_core::type_::Warning::UnusedImportedValue { .. }
                    // | gleam_core::type_::Warning::UnusedLiteral { .. }
                    | gleam_core::type_::Warning::UnusedPrivateFunction { .. }
                    | gleam_core::type_::Warning::UnusedPrivateModuleConstant { .. }
                    | gleam_core::type_::Warning::UnusedType { .. }
                    // | gleam_core::type_::Warning::UnusedValue { .. }
                    | gleam_core::type_::Warning::UnusedVariable { .. },
                ..
            } = warning
            {
                return;
            }
        }
        let buffer_writer = stderr_buffer_writer();
        let mut buffer = buffer_writer.buffer();
        warning.pretty(&mut buffer);
        buffer_writer
            .print(&buffer)
            .expect("Write warning to stderr");
    }
}
