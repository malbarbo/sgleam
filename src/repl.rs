use std::{collections::HashMap, fmt::Write};

use gleam_core::{
    ast::{Definition, Pattern, Statement, TargetedDefinition, UntypedStatement},
    build::Module,
    io::FileSystemWriter,
    Error,
};
use indoc::formatdoc;
use rquickjs::{Array, Context};

use crate::{
    error::{show_error, SgleamError},
    gleam::{compile, get_module, type_to_string, Project},
    javascript::{self, MainFunction},
    parser::{self, ReplItem},
    repl_reader::ReplReader,
    run::get_function,
    swrite, swriteln, GLEAM_MODULES_NAMES,
};

const FNS_REPL: &str = r#"
@external(javascript, "./sgleam_ffi.mjs", "repl_save")
pub fn repl_save(value: a) -> a

@external(javascript, "./sgleam_ffi.mjs", "repl_load")
pub fn repl_load(index: Int) -> a
"#;

pub const QUIT: &str = ":quit";
pub const TYPE: &str = ":type ";

pub fn welcome_message() -> String {
    format!(
        "Welcome to {}.\nType ctrl-d ou \"{QUIT}\" to exit.\n",
        crate::version()
    )
}

#[derive(Clone)]
struct Value {
    index: usize,
    type_: String,
}

#[derive(Clone)]
pub struct Repl {
    user_import: Option<String>,
    imports: Vec<String>,
    consts: Vec<String>,
    types: Vec<String>,
    fns: Vec<(String, Option<String>)>,
    vars: HashMap<String, Value>,
    project: Project,
    context: Context,
    type_: bool,
    iter: usize,
    var_index: usize,
}

enum EntryKind {
    Let(String, String),
    Expr(String),
    Other,
}

impl Repl {
    pub fn new(project: Project, user_module: Option<&Module>) -> Result<Repl, SgleamError> {
        let imports = GLEAM_MODULES_NAMES.iter().map(|s| s.to_string()).collect();
        let fs = project.fs.clone();
        Ok(Repl {
            user_import: user_module.map(import_public_types_and_values),
            imports,
            consts: vec![],
            types: vec![],
            fns: vec![],
            vars: HashMap::new(),
            project,
            context: javascript::create_context(fs, Project::out().into())?,
            type_: false,
            iter: 0,
            var_index: 0,
        })
    }

    pub fn run(&mut self) -> Result<(), SgleamError> {
        let reader = ReplReader::new()?;
        for mut input in reader {
            let line_trim = input.trim();
            if line_trim.is_empty() || line_trim.starts_with("//") {
                continue;
            }

            if line_trim == QUIT {
                break;
            }

            if let Some(expr) = line_trim.strip_prefix(TYPE) {
                self.type_ = true;
                input = expr.into();
            } else {
                self.type_ = false;
            }

            let items = parser::parse_repl(&input).map_err(|error| Error::Parse {
                path: format!("/src/repl{}.gleam", self.iter).into(),
                src: input.clone().into(),
                error,
            });

            let items = match items {
                Err(err) => {
                    self.iter += 1;
                    show_error(&SgleamError::Gleam(err));
                    continue;
                }
                Ok(items) => items,
            };

            // FIXME: avoid this clone
            // We clone self so we can rollback if the execution fail
            let repl = (*self).clone();

            for item in items {
                self.iter += 1;
                let result = match item {
                    ReplItem::ReplDefinition(t) => self.run_definition(t, &input),
                    ReplItem::ReplStatement(s) => self.run_statement(s, &input),
                };

                if let Err(err) = result {
                    show_error(&SgleamError::Gleam(err));
                    *self = repl;
                    break;
                }
            }
        }
        Ok(())
    }

    fn run_code(&mut self, kind: EntryKind) -> Result<(), Error> {
        let mut src = String::new();
        src.push_str(FNS_REPL);
        self.add_imports(&mut src);
        self.add_consts(&mut src);
        self.add_types(&mut src);
        self.add_fns(&mut src);

        match &kind {
            EntryKind::Let(_, expr) => {
                // FIXME: can we generate code that generates better error messagens?
                // Examples of entries that generates poor errors
                // "pub "
                // "let"
                let lets = self.get_lets(&[]);
                src.push_str(&formatdoc! {"
                    pub fn main() {{
                    {lets}
                      io.debug(repl_save({{
                        {expr}
                      }}))
                    }}
                    "
                });
            }
            EntryKind::Expr(expr) => {
                let lets = self.get_lets(&[]);
                src.push_str(&formatdoc! {"
                    pub fn main() {{
                      {lets}
                      io.debug({{
                        {expr}
                      }})
                    }}
                    "
                });
            }
            _ => {
                // main function is not needed
            }
        }

        let iter = self.iter;
        let module_name = format!("repl{iter}");
        let file = format!("{module_name}.gleam");

        // TODO: add an option to show the generated code
        self.project.write_source(&file, &src);

        let result = compile(&mut self.project, true);

        if let Ok(modules) = &result {
            let module = get_module(modules, &module_name).expect("The repl module");
            if let EntryKind::Let(_, _) | EntryKind::Expr(_) = &kind {
                // TODO: change the output for functions (show the function type)
                // TODO: should we show the type of all expressions? See haskell, elm, ocaml, roc.
                if self.type_ {
                    let type_ = get_function(module, "main")
                        .expect("main function")
                        .return_type
                        .clone();
                    println!("{}", type_to_string(module, &type_));
                } else {
                    javascript::run_main(&self.context, &module_name, MainFunction::Main, false);
                }
            } else {
                // Nothing to run, was a definition (type, const, import or fn)
            }

            if let EntryKind::Let(name, _) = &kind {
                if self.try_save_var(name, self.var_index, module) {
                    self.var_index += 1;
                }
            }
        }

        // TODO: Can we remove the file after the compilation?
        self.project
            .fs
            .delete_file(&Project::source().join(file))
            .expect("To delete repl file");

        result.map(|_| ())
    }

    fn run_definition(&mut self, targeted: TargetedDefinition, src: &str) -> Result<(), Error> {
        let start = targeted.definition.location().start as usize;
        let end = targeted.definition.location().end as usize;

        match targeted.definition {
            Definition::Import(_) => {
                let code = String::from(&src[start..end]);
                self.run_import(code)
            }
            Definition::CustomType(t) => {
                let end = t.end_position as usize;
                let code = String::from(&src[start..end]);
                self.run_type(code)
            }
            Definition::TypeAlias(_) => {
                let code = String::from(&src[start..end]);
                self.run_type(code)
            }
            Definition::ModuleConstant(c) => {
                let end = c.value.location().end as usize;
                let code = String::from(&src[start..end]);
                self.run_const(code)
            }
            Definition::Function(f) => {
                let end = f.end_position as usize;
                let mut code = String::from(&src[start..end]);

                let args: Vec<String> = f
                    .arguments
                    .into_iter()
                    .filter_map(|arg| arg.names.get_variable_name().map(String::from))
                    .collect();

                let lets = self.get_lets(&args);

                code.insert_str(
                    f.body.first().location().start as usize - start,
                    &format!("\n  {lets}"),
                );

                let name: Option<String> = f.name.map(|spanned| spanned.1.into());
                self.run_fn(code, name)
            }
        }
    }

    fn run_statement(&mut self, statement: UntypedStatement, src: &str) -> Result<(), Error> {
        let start = statement.location().start as usize;
        let end = statement.location().end as usize;

        match statement {
            Statement::Use(_) => {
                let code = String::from(&src[start..end]);
                self.run_use(code)
            }
            Statement::Expression(_) => {
                let code = String::from(&src[start..end]);
                self.run_expr(code)
            }
            Statement::Assignment(a) => match a.pattern {
                Pattern::Variable { name, .. } => {
                    let end = a.value.location().end as usize;
                    let code = String::from(&src[start..end]);
                    self.run_let(name.into(), code)
                }
                Pattern::Discard { .. } => {
                    let end = a.value.location().end as usize;
                    let code = String::from(&src[start..end]);
                    self.run_expr(code)
                }
                _ => {
                    println!("patterns are not supported in let statements.");
                    Ok(())
                }
            },
        }
    }

    fn run_import(&mut self, _code: String) -> Result<(), Error> {
        println!("imports are not supported.");
        Ok(())
        // TODO: implement import merge
        // import gleam/string.{append}
        // import gleam/string.{inspect}
        // -> import gleam/string.{append, inspect}
        // let new_import = code.trim().strip_prefix("import ").unwrap_or("");
        // self.imports.push(new_import.into());
        // self.run_code(EntryKind::Other)
    }

    fn run_const(&mut self, code: String) -> Result<(), Error> {
        // TODO: improve error message for const redefinition
        self.consts.push(code);
        self.run_code(EntryKind::Other)
    }

    fn run_type(&mut self, code: String) -> Result<(), Error> {
        // TODO: improve error message for type redefinition
        self.types.push(code);
        self.run_code(EntryKind::Other)
    }

    fn run_fn(&mut self, code: String, name: Option<String>) -> Result<(), Error> {
        for (i, (_, other_name)) in self.fns.iter().enumerate() {
            if !other_name.is_none() && *other_name == name {
                let _ = self.fns.remove(i);
                break;
            }
        }

        self.fns.push((code, name));
        self.run_code(EntryKind::Other)
    }

    fn run_let(&mut self, name: String, code: String) -> Result<(), Error> {
        self.run_code(EntryKind::Let(name, code))
    }

    fn run_use(&mut self, _code: String) -> Result<(), Error> {
        println!("use statements are not supported outside blocks.");
        Ok(())
    }

    fn run_expr(&mut self, code: String) -> Result<(), Error> {
        self.run_code(EntryKind::Expr(code))
    }

    fn add_imports(&self, src: &mut String) {
        if let Some(user) = &self.user_import {
            swriteln!(src, "{user}");
        }
        for import in &self.imports {
            swriteln!(src, "import {import}");
        }
    }

    fn add_consts(&self, src: &mut String) {
        for const_ in &self.consts {
            swriteln!(src, "{const_}");
        }
    }

    fn add_types(&self, src: &mut String) {
        for type_ in &self.types {
            swriteln!(src, "{type_}");
        }
    }

    fn add_fns(&self, src: &mut String) {
        for fn_ in &self.fns {
            let (code, _) = fn_;
            swriteln!(src, "{code}");
        }
    }

    fn get_lets(&mut self, exclude: &[String]) -> String {
        let mut lets = String::new();
        for (name, Value { index, type_ }) in &self.vars {
            if !exclude.contains(name) {
                swriteln!(
                    lets,
                    r#"  let {name} = fn () -> {type_} {{ repl_load({index}) }} ()"#
                );
            }
        }
        lets
    }

    fn try_save_var(&mut self, name: &str, index: usize, module: &Module) -> bool {
        if !self.context.with(|ctx| {
            ctx.globals()
                .get::<_, Array>("repl_vars")
                .map(|a| index < a.len())
                .unwrap_or(false)
        }) {
            // the expression crashed and repl_save was not called
            return false;
        }

        let return_type = module
            .ast
            .definitions
            .iter()
            .filter_map(|d| d.main_function())
            .next()
            .expect("The main function")
            .return_type
            .clone();

        self.vars.insert(
            name.into(),
            Value {
                index,
                type_: type_to_string(module, &return_type),
            },
        );

        true
    }
}

fn import_public_types_and_values(module: &Module) -> String {
    let mut import = String::new();
    let name = &module.name;
    swrite!(&mut import, "import {name}.{{");
    for type_ in module.ast.type_info.public_type_names() {
        swrite!(&mut import, "type {type_}, ");
    }
    for value in module.ast.type_info.public_value_names() {
        swrite!(&mut import, "{value}, ");
    }
    import.push('}');
    import
}
