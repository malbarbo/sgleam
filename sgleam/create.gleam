import gleam/int
import sgleam/image.{type Image}
import sgleam/system

pub fn animation(to_image: fn(Int) -> Image, delay: Int) -> Nil {
  animate_loop(to_image, int.max(0, delay), 0)
}

fn animate_loop(to_image: fn(Int) -> Image, delay: Int, frame: Int) {
  system.show_svg(frame |> to_image |> image.to_svg)
  system.sleep(delay)
  animate_loop(to_image, delay, frame + 1)
}
