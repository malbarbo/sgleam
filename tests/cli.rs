use assert_cmd::prelude::*;
use indoc::formatdoc;
use insta::{assert_snapshot, glob};
use sgleam::repl::{welcome_message, QUIT, TYPE};

use std::{
    io::Write,
    process::{Command, Stdio},
};

// FIXME: do not launch sgleam process if not necessary, use the module functions.

#[test]
fn repl_smoke_test() {
    let lit = formatdoc! { r#"
        -2
        13
        4.12
        7.0
        True
        "casa"
        Ok(Nil)"#
    };
    assert_eq!(repl_exec(&lit), lit);
}

#[test]
fn repl_bigint() {
    let lit = "781239812731283189237890123781923";
    assert_eq!(repl_exec(lit), lit);
}

#[test]
fn repl_float_to_string() {
    let lit = "[-1.23, -4.0, 1.234, 3.0, 3.0e21, 1.2e-30, -3.0e56, -1.3e-41]";
    assert_eq!(repl_exec(lit), lit);
}

#[test]
fn repl_constructor_types() {
    let lit = formatdoc! { "
        let a = Ok(10)
        a"
    };
    assert_eq!(repl_exec(&lit), "Ok(10)\nOk(10)");
}

#[test]
fn repl_import() {
    assert_eq!(repl_exec("import something"), "imports are not supported.");
}

#[test]
fn repl_let() {
    assert_eq!(repl_exec("let x = 10\nx + 1"), "10\n11");
}
#[test]
fn repl_let_discard() {
    assert_eq!(repl_exec("let _ = True"), "True");
}

#[test]
fn repl_let_int_pattern() {
    assert_eq!(
        repl_exec("let 10 = 10"),
        "patterns are not supported in let statements."
    );
}

#[test]
fn repl_fn() {
    assert_eq!(repl_exec("fn f(a) { a + 1 }\nf(1)"), "2");
}

#[test]
fn repl_generic_fn() {
    assert_eq!(
        repl_exec("fn keep(_) { True }\nlist.filter([1, 2], keep)"),
        "[1, 2]"
    );
    assert_eq!(
        repl_exec("let keep = fn (_) { True }\nlist.filter([1, 2], keep)"),
        "//fn(a) { ... }\n[1, 2]"
    );
}

#[test]
fn repl_anonymous_fn() {
    assert_eq!(repl_exec("fn () { 1 }"), "//fn() { ... }");
}

#[test]
fn repl_quit() {
    assert_eq!(repl_exec(&format!("{QUIT}\n10")), "");
}

#[test]
fn repl_type() {
    assert_eq!(repl_exec(&format!("{TYPE} 10")), "Int");
    assert_eq!(repl_exec(&format!("{TYPE} int.add")), "fn(Int, Int) -> Int");
    assert_eq!(
        repl_exec(&format!("{TYPE} list.filter_map")),
        "fn(List(b), fn(b) -> Result(c, d)) -> List(c)"
    );
    // :type does not evaluate
    assert_eq!(
        repl_exec(&format!("{TYPE} io.debug(Ok(1))")),
        "Result(Int, b)", // without the io.debug side effect
    );
    // TODO: check that :type let x = 10 does not create x
}

#[test]
fn repl_type_module() {
    assert_eq!(
        repl_exec(&format!("type List {{}}\n{TYPE} list.map")),
        "fn(gleam.List(b), fn(b) -> c) -> gleam.List(c)"
    );
}

#[test]
fn repl_user_module_import() {
    let input = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs/user.gleam");
    assert_eq!(
        run_sgleam_cmd_stdout(
            &["-q", "-i", input],
            Some(
                "one
                  two()
                  let _: Three = Num3"
            )
        ),
        "1\n2\nNum3\n"
    );
}

#[test]
fn format_stdin() {
    assert_eq!(
        run_sgleam_cmd_stdout(
            &["-f"],
            Some(&formatdoc! {r#"
            import gleam / io.{{ debug , }}
            fn main() {{
               debug("Hello world!" )
            }}
            "#}),
        ),
        formatdoc! {r#"
        import gleam/io.{{debug}}

        fn main() {{
          debug("Hello world!")
        }}
        "#}
    )
}

#[test]
fn repl_welcome_message() {
    assert_eq!(run_sgleam_cmd_stdout(&[], None), welcome_message())
}

fn repl_exec(s: &str) -> String {
    run_sgleam_cmd_stdout(&["-q"], Some(s))
        .strip_suffix('\n')
        .unwrap_or("")
        .into()
}

#[test]
fn run_tests() {
    glob!("inputs/check*.gleam", |path| {
        let (out, err) = run_sgleam_cmd(
            &["-t", path.as_os_str().to_str().expect("a valid path")],
            None,
        );
        assert_snapshot!(formatdoc! {"
            STDOUT
            {out}
            STDERR
            {err}"
        });
    });
}

#[test]
fn run_file() {
    glob!("inputs/*.gleam", |path| {
        let (out, err) = run_sgleam_cmd(&[&path.to_string_lossy().to_string()], None);
        assert_snapshot!(formatdoc! {"
            STDOUT
            {out}
            STDERR
            {err}"
        });
    });
}

#[test]
fn smain_list_string() {
    let input = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/tests/inputs/smain_list_string.gleam"
    );
    let (out, err) = run_sgleam_cmd(
        &[input],
        Some(&formatdoc! {
            "
            An example
            with
            three lines"
        }),
    );
    assert_snapshot!(formatdoc! {"
        STDOUT
        {out}
        STDERR
        {err}"
    });
}

#[test]
fn smain_string() {
    let input = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/tests/inputs/smain_string.gleam"
    );
    let (out, err) = run_sgleam_cmd(&[input], Some("hello\nworld"));
    assert_snapshot!(formatdoc! {"
        STDOUT
        {out}
        STDERR
        {err}"
    });
}

fn run_sgleam_cmd_stdout(args: &[&str], input: Option<&str>) -> String {
    run_sgleam_cmd(args, input).0
}

// FIXME: this seams too complicated
fn run_sgleam_cmd(args: &[&str], input: Option<&str>) -> (String, String) {
    let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();

    let mut child = cmd
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .args(args)
        .spawn()
        .expect("Spawn child process");

    if let Some(input) = input.map(|input| format!("{input}\n")) {
        let mut stdin = child.stdin.take().expect("Open stdin");
        std::thread::spawn(move || {
            stdin.write_all(input.as_bytes()).expect("Write to stdin");
        });
    }

    let result = child.wait_with_output().unwrap();

    // assert!(result.status.success());

    (
        String::from_utf8_lossy(&result.stdout).into_owned(),
        String::from_utf8_lossy(&result.stderr).into_owned(),
    )
}
