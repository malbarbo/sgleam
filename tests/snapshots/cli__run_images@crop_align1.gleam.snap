---
source: tests/cli.rs
expression: "format!(\"{out}\")"
input_file: tests/images/crop_align1.gleam
---
<svg width="40.0" height="40.0" xmlns="http://www.w3.org/2000/svg">
  <defs><clipPath id="clip0" ><rect x="0.0" y="0.0" width="40.0" height="40.0" transform="rotate(0.0 20.0 20.0)" /></clipPath></defs>
  <g clip-path="url(#clip0)" >
    <ellipse cx="40.0" cy="40.0" rx="40.0" ry="40.0" transform="rotate(0.0 40.0 40.0)" fill="rgba(210, 105, 30, 1.0)" />
  </g>
</svg>
